### ISSUE TITLE
A swift object with a leading '/' character in its name is not properly handled.

### ISSUE THREAD

**Issue by cazino** [scality/ScalitySproxydSwift#100]
According to the code of the swift functional test suite, a leading '/' is meaningfull in an object name.

See here https://github.com/openstack/swift/blob/master/test/functional/tests.py#L639 and here https://github.com/openstack/swift/blob/master/test/functional/tests.py#L656. These are supposed to be 2 different objects.

The driver make no difference between these two objects, a '/x' object and 'x' are the same.

This leads to bug like this one : https://37.187.159.67:5443/job/swift-functional-tests/DEVSTACK_BRANCH=stable%2Fjuno,host=ubuntu14-large/103/testReport/junit/test.functional.tests/TestContainerUTF8/testContainerExistenceCachingProblem/ where a 'dir1/' object can't be found (HTTP 404) because a '/dir1/' object was deleted before.

The reason is here : https://github.com/scality/ScalitySproxydSwift/blob/master/swift_scality_backend/diskfile.py#L215
We send HTTP request to sproxyd with double '/' in the url. I suppose that a double '/' is silently transformed into single '/' by most http server; at least I can't find any double '/' in sproxyd logs.

**Comment by JordanP** [comment]
+1. To test :
```
curl -v -XPUT -H "Expect:" http://127.0.0.1:81/proxy/chord_path//test --data-binary @/etc/hosts
curl -v -XGET -H "Expect:" http://127.0.0.1:81/proxy/chord_path/test
```
Indeed the problem is because we also join account, container and object name with '/', so two '/' can be present.

**Comment by cazino** [comment]
Fix proposal : we need somehow to transform the object name before injecting it in a URL.

Option 1 : To stay as much backward compatible as possible we might want to transform only leading slashes, if present, and transform them into 2 characters to avoid clashes with substituted character. We could use this mapping for example : '/' become '--'.

This only fixes leading slash bug.  The bug can also occur with successive '/' in object name, for example 'dir1//file1'.

Option 2 : If we want to handle that as well we can transform every '/' occurrence and therefore hurt more backward compatibility. Do we want that ? 

Option 3 : Alternatively we could transform only the second '/' occurrence in case of of two successive occurrence. Backward compatibility is more preserved that way but with the cost of a regxep computation each time an object is accessed, with does not sound as a good idea.

I vote option 1.

**Comment by JordanP** [comment]
Option 4 (if it works): use `urllib.quote_plus` (instead of `urllib.quote`) which encodes '/' into '%2F'

**Comment by cazino** [comment]
In my opinion the problem with option 4 is that the object name will be quoted two times, one time by the swift_scality_backend, one time by the  scality_sproxyd_client.

**Comment by JordanP** [comment]
I meant using urllib.quote_plus (**instead** of urllib.quote), in https://github.com/scality/scality-sproxyd-client/blob/master/scality_sproxyd_client/sproxyd_client.py#L200

We should not quote anything in swift_scality_backend.

**Comment by cazino** [comment]
Yes agreed.
But it completely break backward compatibility for every object (they all have at least two  '/' in their name : 'account/container/objectname')

**Comment by cazino** [comment]
Regarding option 1, 2 and 3 : transforming a '/' character in a double character '--' does not do the trick, it is not better than transforming to any single character, it does not avoid object name collision.

**Comment by NicolasT** [comment]
@JordanP You mentioned some Apache configuration option. Could you summarize your findings here?

**Comment by JordanP** [comment]
Yup.

First I re-read http://svn.xe15.com/trac/ticket/12002

The interesting Apache configuration parameter is `AllowEncodedSlashes On|Off|NoDecode`. You can put that in a Vhost section, default value is `Off`. By default, Apache2 will return a 404, if it sees the '%2F'  sequence in a URL.

If value is `On`, Apache will first decode '%2F' to '/' and pass the request to the CGI program (Sproxyd). It means, as a user, you can send a request to Apache with `/` or `%2F` in the request path,  alternatively. You end up with the same result.

If value is `NoDecode`, Apache won't decode `%2F`, so `/` is different from `%2F`.
 
To fix the overall issue I see 3 parameters we can tweak :
1. AllowEncodedSlashes in Apache2.conf
2. What is the separator in URLs between `account`, `container`, and `object`. Currently it is `/`  (see https://github.com/scality/ScalitySproxydSwift/blob/master/swift_scality_backend/diskfile.py#L215). It could be `%2F` (which will or will not break compatibility depending on 1.)
3. Where/what we should encode using `urllib.quote()`. Currently with encode the entire request path, because we call `urllib.quote()` in the Sproxyd Client which is not aware of the different components (`account`, `container`, and `object` of the URL). We could encode only the object name, but we would have to do the encoding in the `diskfile`.

Remember we need to encode `/` into `%2F` somewhere because otherwise, Apache2 will make Sproxyd believe that an object named `my//object` is the same as `my/object`.

**Comment by NicolasT** [comment]
As a separator, which is now `/`, it could be `%2F`, but it could also be something completely different, no? Isn't that a (non-backwards-compatible) way out?

**Comment by JordanP** [comment]
If you use something different than / or %2F, you are sure to lose track of every object stored in the previous versions of the driver. Maybe there's no way around and we will have to do this.

But, even if we change the separator to something different, we would not be able to distinguish named "my//obj" from an object named "my/obj". (I am not sure two forward slashes are allowed in object name though one slash is valid, to delimit pseudo folders http://docs.rackspace.com/files/api/v1/cf-devguide/content/Pseudo-Hierarchical_Folders_Directories-d1e1580.html )

**Comment by JordanP** [comment]
According to John Dickinson, 
```jordanP: /v1/AUTH_foo/barcontainer/x/////y   <-- totally valid path in swift```
